File: src/lib/ai/gemini.ts
import { GoogleGenerativeAI } from "@google/generative-ai";
import type { RunConfig, WebsiteAnalysis, ExecutiveReport, Insights, CompetitorAnalysis, RoastRegion, AutoFix } from "@/lib/types";
import { mockAnalyze, buildReport } from "@/lib/data/mock-engine";
import { sleep } from "@/lib/utils";
import { groqEnabled, groqJSON, groqVisionEnabled, groqVisionJSON } from "@/lib/ai/groq";

// ---------------------------------------------------------------------------
// Gemini client with graceful degradation.
//
// Everything here returns a valid result even when GEMINI_API_KEY is unset or
// the API errors — it falls back to the deterministic mock engine. AI is used
// where it adds the most value (understanding crawled site content and writing
// the executive narrative); per-persona scoring stays in the fast, grounded
// engine so a 200-customer swarm doesn't fan out into 200 fragile LLM calls.
// ---------------------------------------------------------------------------

// Primary model first, then resilient fallbacks. When the flagship model is
// overloaded (HTTP 503 "high demand"), we transparently retry on an alternate
// so real AI analysis survives transient spikes instead of dropping to mock.
const PRIMARY = process.env.GEMINI_MODEL || "gemini-2.5-flash";
// flash-lite has a separate quota and stays available when flash is rate-limited;
// 1.5-flash is 404 for current keys so it's intentionally NOT in the chain.
const MODELS = [...new Set([PRIMARY, "gemini-2.5-flash-lite", "gemini-2.0-flash"])];
const KEY = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;

export function isGeminiEnabled(): boolean {
  return Boolean(KEY);
}

/** Visual UI Roast works if EITHER Gemini Vision or Groq Vision is available. */
export function isVisionEnabled(): boolean {
  return isGeminiEnabled() || groqVisionEnabled();
}

export function engineName(): "gemini" | "mock" {
  // "gemini" here means "a real LLM is available" (Groq or Gemini) vs "mock".
  return isGeminiEnabled() || groqEnabled() ? "gemini" : "mock";
}

let client: GoogleGenerativeAI | null = null;
function getClient(): GoogleGenerativeAI | null {
  if (!KEY) return null;
  if (!client) client = new GoogleGenerativeAI(KEY);
  return client;
}

async function withTimeout<T>(p: Promise<T>, ms: number): Promise<T> {
  return await Promise.race([
    p,
    new Promise<T>((_, rej) => setTimeout(() => rej(new Error("gemini timeout")), ms)),
  ]);
}

/**
 * Call Gemini expecting strict JSON. Retries transient overloads (503/429/
 * timeout) and falls back across alternate models before giving up. Returns
 * null only when every model/attempt fails (then the caller uses the mock).
 */
export async function generateJSON<T>(prompt: string, timeoutMs = 30000): Promise<T | null> {
  // Prefer Groq for text generation (fast + generous free quota); fall back to
  // Gemini if Groq is unavailable/fails; the caller falls back to mock if both fail.
  if (groqEnabled()) {
    const g = await groqJSON<T>(prompt, Math.min(timeoutMs, 22000));
    if (g) return g;
  }

  const c = getClient();
  if (!c) return null;

  for (const modelName of MODELS) {
    for (let attempt = 0; attempt < 2; attempt++) {
      try {
        const model = c.getGenerativeModel({
          model: modelName,
          generationConfig: { responseMimeType: "application/json", temperature: 0.7 },
        });
        const res = await withTimeout(model.generateContent(prompt), timeoutMs);
        return JSON.parse(stripFences(res.response.text())) as T;
      } catch (err) {
        const msg = ((err as Error).message || "").slice(0, 140);
        console.warn(`[gemini] ${modelName} attempt ${attempt + 1} failed: ${msg}`);
        // 429 (rate limit/quota) won't clear in ~1s -> jump to the next model now
        if (/429|quota|rate limit|too many|resource exhausted/i.test(msg)) break;
        const transient = /503|overload|high demand|timeout|unavailable|fetch failed|ECONN|ETIMEDOUT/i.test(msg);
        if (transient && attempt === 0) {
          await sleep(900);
          continue; // retry same model once
        }
        break; // give up on this model, try the next
      }
    }
  }
  console.warn("[gemini] all models/attempts failed -> using mock engine");
  return null;
}

function stripFences(s: string): string {
  const t = s.trim();
  if (t.startsWith("```")) {
    return t.replace(/^```(?:json)?\s*/i, "").replace(/```\s*$/i, "");
  }
  return t;
}

// ---------------------------------------------------------------------------
// AI-enhanced website analysis (from crawled content)
// ---------------------------------------------------------------------------
export async function aiAnalyze(
  config: RunConfig,
  crawledText: string | null,
  crawlSource: "firecrawl" | "fetch" | "none" = "none",
): Promise<WebsiteAnalysis> {
  const fallback = mockAnalyze(config); // fallback.source === "mock"
  if (!isGeminiEnabled() || !crawledText) return fallback;

  const prompt = `You are a B2B website analyst. Analyze the following website content and return STRICT JSON only.

URL: ${config.url}
CONTENT (may be truncated):
"""
${crawledText.slice(0, 12000)}
"""

Return JSON with this exact shape:
{
  "title": string,
  "tagline": string,
  "category": string,
  "valueProps": string[3],
  "ctas": string[],
  "navItems": string[],
  "pricingTiers": [{"name": string, "price": string, "period": string, "highlights": string[]}],
  "faqs": [{"q": string, "a": string}],
  "trustSignals": string[],
  "missingTrustSignals": string[],
  "onboardingSteps": string[],
  "detectedFeatures": string[],
  "targetAudience": string,
  "toneOfVoice": string,
  "contentScore": number  // 0-100, how clearly the site communicates value, pricing, trust
}
Be accurate to the content. If pricing or security info is absent, reflect that (empty arrays, low contentScore, list them in missingTrustSignals).`;

  const json = await generateJSON<Partial<WebsiteAnalysis>>(prompt);
  if (!json || !json.title) return fallback;

  // Merge AI output over a safe fallback so missing fields never break the UI.
  return {
    ...fallback,
    ...json,
    url: config.url,
    pricingTiers: json.pricingTiers?.length ? json.pricingTiers : fallback.pricingTiers,
    faqs: json.faqs ?? fallback.faqs,
    valueProps: json.valueProps?.length ? json.valueProps : fallback.valueProps,
    detectedFeatures: json.detectedFeatures?.length ? json.detectedFeatures : fallback.detectedFeatures,
    trustSignals: json.trustSignals ?? fallback.trustSignals,
    missingTrustSignals: json.missingTrustSignals ?? fallback.missingTrustSignals,
    contentScore: typeof json.contentScore === "number" ? json.contentScore : fallback.contentScore,
    // honest provenance: real AI analysis of a real crawl, or plain fetch
    source: crawlSource === "none" ? "fetch" : crawlSource,
  } as WebsiteAnalysis;
}

// ---------------------------------------------------------------------------
// AI-enhanced executive report narrative
// ---------------------------------------------------------------------------
export async function aiReport(
  analysis: WebsiteAnalysis,
  insights: Insights,
  competitor?: CompetitorAnalysis,
): Promise<ExecutiveReport> {
  const fallback = buildReport(analysis, insights, competitor);
  if (!isGeminiEnabled()) return fallback;

  const prompt = `You are a principal growth strategist writing an executive report for the business "${analysis.title}" (${analysis.category}).
Use ONLY the data below. Return STRICT JSON matching the schema. Be specific, quantitative, and board-ready.

DATA:
${JSON.stringify(
    {
      avgPurchaseProbability: insights.avgPurchaseProbability,
      avgConfusion: insights.avgConfusion,
      conversionRiskScore: insights.conversionRiskScore,
      verdictBreakdown: insights.verdictBreakdown,
      roleBreakdown: insights.roleBreakdown,
      revenueLeaks: insights.revenueLeaks,
      churnRisks: insights.churnRisks,
      unansweredQuestions: insights.salesObjections.filter((o) => !o.answeredOnSite).map((o) => o.question),
    },
    null,
    2,
  )}

Schema:
{
  "executiveSummary": string,  // 3-4 sentences
  "keyFindings": string[4],
  "conversionRisks": string[],
  "churnRisks": string[],
  "revenueLeaks": string[],
  "customerQuestions": string[],
  "competitorNotes": string[],
  "recommendations": [{"title": string, "detail": string, "effort": "Low"|"Medium"|"High", "impact": "Low"|"Medium"|"High"}],
  "projectedUplift": string
}`;

  const json = await generateJSON<ExecutiveReport>(prompt, 22000);
  if (!json || !json.executiveSummary) return fallback;
  return {
    ...fallback,
    ...json,
    recommendations: json.recommendations?.length ? json.recommendations : fallback.recommendations,
    customerQuestions: json.customerQuestions?.length ? json.customerQuestions : fallback.customerQuestions,
  };
}

// ---------------------------------------------------------------------------
// AI "live thought" for the war-room stream (optional flavor; falls back to "")
// ---------------------------------------------------------------------------
export async function aiThought(agent: string, context: string): Promise<string | null> {
  if (!isGeminiEnabled()) return null;
  const json = await generateJSON<{ thought: string }>(
    `You are the "${agent}" inside a customer-simulation engine. In ONE punchy sentence (max 18 words), narrate what you just observed. Context: ${context}. Return JSON {"thought": string}.`,
    8000,
  );
  return json?.thought ?? null;
}

// ---------------------------------------------------------------------------
// Visual Roast — Gemini Vision over a real screenshot.
// Returns null on any failure (caller surfaces an honest error; never mock).
// ---------------------------------------------------------------------------
export interface VisionRoastData {
  roast: string;
  clarityScore: number;
  firstLook?: { x: number; y: number };
  regions: RoastRegion[];
  engine?: "gemini" | "groq";
}

export async function aiVisionRoast(screenshotUrl: string): Promise<VisionRoastData | null> {
  // fetch the screenshot bytes -> base64 (shared by both vision engines)
  let b64 = "";
  let mime = "image/png";
  try {
    const r = await withTimeout(fetch(screenshotUrl), 30000);
    if (!r.ok) return null;
    mime = r.headers.get("content-type") || "image/png";
    const ab = await r.arrayBuffer();
    b64 = Buffer.from(ab).toString("base64");
  } catch {
    return null;
  }

  const prompt = `You are a first-time, slightly impatient potential customer landing on this web page for the first time. Be brutally honest but specific.

Return STRICT JSON only:
{
  "roast": string,            // <= 2 sentences, blunt first-impression of the page
  "clarityScore": number,     // 0-100, how clearly the page communicates value + next action
  "firstLook": { "x": number, "y": number },  // where your eye lands FIRST, normalized 0-1000
  "regions": [                // 4 to 7 items: spots that cause confusion / friction / lost attention
    {
      "label": string,        // short name, e.g. "Buried CTA", "Wall of text", "Unclear pricing"
      "why": string,          // one sentence: why it confuses or loses you
      "severity": "high" | "medium" | "low",
      "box": [ymin, xmin, ymax, xmax]   // bounding box, integers normalized to 0-1000 (Gemini convention)
    }
  ]
}
Focus on real, visible problems in THIS screenshot (hero, CTA, navigation, pricing, trust, readability). Coordinates must be within 0-1000.`;

  // Engine 1: Gemini Vision (preferred when a key is configured). One fast
  // attempt per model — when Gemini is slow/over quota we want to fail over to
  // the Groq fallback quickly rather than burn ~2 min retrying a busy model.
  const c = getClient();
  if (c) {
    for (const modelName of MODELS) {
      try {
        const model = c.getGenerativeModel({
          model: modelName,
          generationConfig: { responseMimeType: "application/json", temperature: 0.4 },
        });
        const res = await withTimeout(
          model.generateContent([{ text: prompt }, { inlineData: { mimeType: mime, data: b64 } }]),
          18000,
        );
        const parsed = JSON.parse(stripFences(res.response.text())) as VisionRoastData;
        if (parsed && Array.isArray(parsed.regions)) return { ...parsed, engine: "gemini" };
      } catch (err) {
        const msg = ((err as Error).message || "").slice(0, 140);
        console.warn(`[gemini-vision] ${modelName} failed: ${msg}`);
        if (/429|quota|rate limit|too many|resource exhausted/i.test(msg)) break; // straight to Groq
      }
    }
  }

  // Engine 2: Groq Vision fallback — keeps the Roast working when Gemini is out
  // of quota or busy (the exact "Vision analysis failed (model busy)" case).
  if (groqVisionEnabled()) {
    const parsed = await groqVisionJSON<VisionRoastData>(prompt, `data:${mime};base64,${b64}`, 30000);
    if (parsed && Array.isArray(parsed.regions)) return { ...parsed, engine: "groq" };
  }

  return null;
}

// ---------------------------------------------------------------------------
// Auto-Fix — Gemini generates optimized copy/markup for a detected problem.
// ---------------------------------------------------------------------------
export async function aiAutoFix(
  problem: { title: string; cause?: string; fix?: string },
  context: { title?: string; category?: string; tone?: string; audience?: string },
): Promise<AutoFix | null> {
  if (!isGeminiEnabled()) return null;
  const prompt = `You are a senior conversion copywriter + front-end engineer. A customer-simulation found this problem on the website "${context.title ?? "the site"}" (${context.category ?? "web product"}):

PROBLEM: ${problem.title}
${problem.cause ? `CAUSE: ${problem.cause}` : ""}
${problem.fix ? `SUGGESTED DIRECTION: ${problem.fix}` : ""}
BRAND TONE: ${context.tone ?? "clear, confident, friendly"}
TARGET AUDIENCE: ${context.audience ?? "the site's prospective buyers"}

Generate concrete, paste-ready fixes that directly overcome this problem. Return STRICT JSON:
{
  "problem": string,        // restate the problem in one line
  "rationale": string,      // 1-2 sentences: why these fixes convert better
  "variants": [             // 2 to 3 variants
    {
      "kind": string,       // e.g. "Hero headline", "Pricing blurb", "FAQ entry", "Trust strip"
      "heading": string,    // short label for this variant
      "copy": string,       // the actual ready-to-use copy
      "html": string        // a small self-contained Tailwind HTML snippet implementing it (no <script>)
    }
  ]
}
Keep copy specific and benefit-led; no lorem ipsum; no scripts in html.`;
  const json = await generateJSON<AutoFix>(prompt, 22000);
  if (!json || !Array.isArray(json.variants) || !json.variants.length) return null;
  return json;
}

File: agents-py/gemini.py
"""
Ghost Customer AI — Gemini wrapper.

Thin, defensive wrapper around `google-generativeai` (model: gemini-2.5-flash).

Core contract
-------------
* `is_enabled()`            -> True only if GEMINI_API_KEY is set AND the SDK
                               imported cleanly.
* `generate_json(prompt)`   -> parsed `dict` (JSON mode) or `None` on any
                               error/timeout. NEVER raises to the caller.
* `ai_analyze(text, url)`   -> WebsiteAnalysis (Gemini if possible, else mock).
* `ai_report(...)`          -> ExecutiveReport (Gemini if possible, else mock).

CRITICAL DESIGN RULE: the engine must run with NO keys. Every AI helper falls
back to the deterministic mock so the pipeline always completes.
"""

from __future__ import annotations

import json
import os
from typing import List, Optional

from dotenv import load_dotenv

import mock
from models import ExecutiveReport, Insights, WebsiteAnalysis

load_dotenv()

GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-2.5-flash")
_API_KEY = os.getenv("GEMINI_API_KEY", "").strip()

# Soft import — the SDK may be absent in mock-only deployments.
try:  # pragma: no cover - import guard
    import google.generativeai as genai

    _SDK_OK = True
except Exception:  # noqa: BLE001
    genai = None  # type: ignore[assignment]
    _SDK_OK = False

_configured = False


def is_enabled() -> bool:
    """True if we have both a key and a working SDK."""
    return bool(_API_KEY) and _SDK_OK


def _ensure_configured() -> bool:
    global _configured
    if not is_enabled():
        return False
    if not _configured:
        try:
            genai.configure(api_key=_API_KEY)  # type: ignore[union-attr]
            _configured = True
        except Exception:  # noqa: BLE001
            return False
    return True


def generate_json(prompt: str, *, timeout: float = 45.0) -> Optional[dict]:
    """
    Run a JSON-mode generation and return a parsed dict, or None on any failure.

    We request `response_mime_type=application/json` so the model returns a raw
    JSON object. We still parse defensively in case of stray markdown fences.
    """
    if not _ensure_configured():
        return None
    try:
        model = genai.GenerativeModel(  # type: ignore[union-attr]
            GEMINI_MODEL,
            generation_config={
                "response_mime_type": "application/json",
                "temperature": 0.7,
            },
        )
        resp = model.generate_content(
            prompt,
            request_options={"timeout": timeout},
        )
        raw = (getattr(resp, "text", None) or "").strip()
        if not raw:
            return None
        return _safe_json(raw)
    except Exception:  # noqa: BLE001  (timeouts, quota, parse, network — all -> None)
        return None


def _safe_json(raw: str) -> Optional[dict]:
    """Parse JSON, tolerating ```json fences or leading/trailing prose."""
    try:
        return json.loads(raw)
    except Exception:  # noqa: BLE001
        # Strip code fences and retry on the first {...} block.
        cleaned = raw.strip()
        if cleaned.startswith("```"):
            cleaned = cleaned.split("```", 2)
            cleaned = cleaned[1] if len(cleaned) > 1 else raw
            if cleaned.lstrip().lower().startswith("json"):
                cleaned = cleaned.lstrip()[4:]
        start = cleaned.find("{")
        end = cleaned.rfind("}")
        if start != -1 and end != -1 and end > start:
            try:
                return json.loads(cleaned[start : end + 1])
            except Exception:  # noqa: BLE001
                return None
        return None


# ---------------------------------------------------------------------------
# High-level helpers — Gemini if possible, deterministic mock otherwise.
# ---------------------------------------------------------------------------
def ai_analyze(crawled_text: Optional[str], url: str, source: str = "mock") -> WebsiteAnalysis:
    """
    Turn raw crawled page text into a structured WebsiteAnalysis.

    If Gemini is unavailable OR returns something unusable, we fall back to the
    deterministic mock (stamped with the true `source` so the UI still shows
    whether the text was crawled).
    """
    if not crawled_text or not is_enabled():
        return mock.mock_analysis(url, source=source)

    prompt = _ANALYZE_PROMPT.format(
        url=url, text=crawled_text[:18000]  # keep well under token limits
    )
    data = generate_json(prompt)
    if not data:
        return mock.mock_analysis(url, source=source)

    # Validate against the schema; on any mismatch, fall back gracefully but
    # keep whatever good fields we can by merging onto a mock skeleton.
    try:
        data.setdefault("url", url)
        data["source"] = source if source in ("firecrawl", "fetch") else "mock"
        return WebsiteAnalysis.model_validate(data)
    except Exception:  # noqa: BLE001
        skeleton = mock.mock_analysis(url, source=source).model_dump()
        skeleton.update({k: v for k, v in data.items() if k in skeleton})
        try:
            return WebsiteAnalysis.model_validate(skeleton)
        except Exception:  # noqa: BLE001
            return mock.mock_analysis(url, source=source)


def ai_report(
    url: str,
    analysis: WebsiteAnalysis,
    insights: Insights,
    competitor=None,
) -> ExecutiveReport:
    """
    Produce a narrative ExecutiveReport. Gemini writes the prose when enabled;
    otherwise we use the deterministic mock report.
    """
    if not is_enabled():
        return mock.mock_report(url, analysis, insights, competitor)

    prompt = _REPORT_PROMPT.format(
        url=url,
        analysis=json.dumps(analysis.model_dump(), ensure_ascii=False)[:6000],
        insights=json.dumps(insights.model_dump(), ensure_ascii=False)[:8000],
    )
    data = generate_json(prompt)
    if not data:
        return mock.mock_report(url, analysis, insights, competitor)
    try:
        # Backfill competitor notes from the deterministic analysis if the
        # model omitted them but we have competitor data.
        if competitor and not data.get("competitorNotes"):
            data["competitorNotes"] = [
                f"Head-to-head winner: {competitor.winner} — {competitor.reason}"
            ]
        return ExecutiveReport.model_validate(data)
    except Exception:  # noqa: BLE001
        return mock.mock_report(url, analysis, insights, competitor)


# ---------------------------------------------------------------------------
# Prompts (kept terse + schema-locked; JSON mode does the heavy lifting)
# ---------------------------------------------------------------------------
_ANALYZE_PROMPT = """You are the Website Analyzer agent for "Ghost Customer AI".
Analyze the crawled landing page below and return ONLY a JSON object matching
this exact schema (camelCase keys, no extra keys):

{{
  "url": string,
  "title": string,
  "tagline": string,
  "category": string,
  "valueProps": string[],
  "ctas": string[],
  "navItems": string[],
  "pricingTiers": [{{"name": string, "price": string, "period": string|null, "highlights": string[]}}],
  "faqs": [{{"q": string, "a": string}}],
  "trustSignals": string[],
  "missingTrustSignals": string[],
  "onboardingSteps": string[],
  "detectedFeatures": string[],
  "targetAudience": string,
  "toneOfVoice": string,
  "contentScore": number,
  "source": "mock"
}}

contentScore is 0-100 overall clarity. missingTrustSignals are trust elements a
buyer would expect but that are absent (e.g. SOC 2, SLA, testimonials).

URL: {url}

CRAWLED PAGE TEXT:
{text}
"""

_REPORT_PROMPT = """You are the Report Generator agent for "Ghost Customer AI".
Write a sharp, executive-ready report based on the website analysis and the
aggregated insights. Return ONLY a JSON object matching this exact schema:

{{
  "executiveSummary": string,
  "keyFindings": string[],
  "conversionRisks": string[],
  "churnRisks": string[],
  "revenueLeaks": string[],
  "customerQuestions": string[],
  "competitorNotes": string[],
  "recommendations": [{{"title": string, "detail": string, "effort": "Low"|"Medium"|"High", "impact": "Low"|"Medium"|"High"}}],
  "projectedUplift": string
}}

Be specific and quantitative. Tie every recommendation to a revenue leak or
churn risk. Tone: confident, concise, board-room.

URL: {url}

WEBSITE ANALYSIS (JSON):
{analysis}

INSIGHTS (JSON):
{insights}
"""
